File: src/services/es.ts
/**
 * Typed bridge to the Rust Elasticsearch command layer.
 *
 * The WebView NEVER talks to Elasticsearch directly — every call here invokes a
 * Rust command (reqwest transport, credentials resolved from the macOS Keychain
 * on the Rust side). Command names + payload shapes mirror src-tauri/src/commands.
 * tauri-specta also emits src/bindings.ts as the canonical typed contract.
 */

import { invoke, isTauri } from "@tauri-apps/api/core";
import type { Connection, Dsl, EsError, Field, Health, Hit, SearchResult } from "../lib/types";

/** A raw, transport-level failure (could not reach the cluster). */
export class EsTransportError extends Error {}

export interface IndexInfo {
  name: string;
  docs: number;
  health?: Health;
}

export interface ClusterInfo {
  health: Health;
  version: string;
  nodes: number;
}

export interface TestResult {
  ok: boolean;
  tookMs: number;
  version?: string;
  nodes?: number;
  health?: Health;
  message?: string;
}

/** Connection metadata passed to Rust (the secret is resolved server-side). */
interface ConnInput {
  id: string;
  host: string;
  auth: string;
  user: string;
}

const connInput = (c: Connection): ConnInput => ({
  id: c.id,
  host: c.host,
  auth: c.auth,
  user: c.user,
});

export const hasTauri = (): boolean => isTauri();

async function call<T>(cmd: string, args: Record<string, unknown>): Promise<T> {
  try {
    return await invoke<T>(cmd, args);
  } catch (e) {
    // Rust returns ES-level failures inside the result; transport/keychain
    // failures throw a string here. Re-wrap so callers can branch.
    throw new EsTransportError(typeof e === "string" ? e : (e as Error)?.message ?? String(e));
  }
}

export async function esSearch(conn: Connection, index: string, dsl: Dsl): Promise<SearchResult> {
  return call<SearchResult>("es_search", { conn: connInput(conn), index, dsl });
}

export async function esGetDoc(conn: Connection, index: string, id: string): Promise<Hit | null> {
  return call<Hit | null>("es_get_doc", { conn: connInput(conn), index, id });
}

export async function esListIndices(conn: Connection): Promise<IndexInfo[]> {
  return call<IndexInfo[]>("es_list_indices", { conn: connInput(conn) });
}

export async function esClusterHealth(conn: Connection): Promise<ClusterInfo> {
  return call<ClusterInfo>("es_cluster_health", { conn: connInput(conn) });
}

export async function esFieldCaps(conn: Connection, index: string): Promise<Field[]> {
  return call<Field[]>("es_field_caps", { conn: connInput(conn), index });
}

export async function esTestConnection(conn: Connection): Promise<TestResult> {
  return call<TestResult>("es_test_connection", { conn: connInput(conn) });
}

// ---- Keychain (secrets stay on the Rust side) ------------------------------

export async function setCredential(connId: string, secret: string): Promise<void> {
  if (!isTauri()) return;
  await call<null>("set_credential", { connId, secret });
}

export async function deleteCredential(connId: string): Promise<void> {
  if (!isTauri()) return;
  await call<null>("delete_credential", { connId });
}

/** Narrow an arbitrary error into our readable EsError, if it is one. */
export function asEsError(e: unknown): EsError | null {
  if (e && typeof e === "object" && "reason" in e && "type" in e) {
    return e as EsError;
  }
  return null;
}

File: src-tauri/src/commands/es.rs
//! Thin Tauri command wrappers over the ES transport. All Elasticsearch I/O
//! happens here on the Rust side; the WebView only ever invokes these.

use serde_json::Value;

use crate::es::client;
use crate::es::model::*;

#[tauri::command]
#[specta::specta]
pub async fn es_search(conn: ConnInput, index: String, dsl: Value) -> Result<SearchResult, String> {
    client::search(&conn, &index, dsl).await
}

#[tauri::command]
#[specta::specta]
pub async fn es_get_doc(conn: ConnInput, index: String, id: String) -> Result<Option<Hit>, String> {
    client::get_doc(&conn, &index, &id).await
}

#[tauri::command]
#[specta::specta]
pub async fn es_list_indices(conn: ConnInput) -> Result<Vec<IndexInfo>, String> {
    client::list_indices(&conn).await
}

#[tauri::command]
#[specta::specta]
pub async fn es_cluster_health(conn: ConnInput) -> Result<ClusterInfo, String> {
    client::cluster_health(&conn).await
}

#[tauri::command]
#[specta::specta]
pub async fn es_field_caps(conn: ConnInput, index: String) -> Result<Vec<Field>, String> {
    client::field_caps(&conn, &index).await
}

#[tauri::command]
#[specta::specta]
pub async fn es_test_connection(conn: ConnInput) -> Result<TestResult, String> {
    client::test_connection(&conn).await
}
